Tools: Enable Unit Testing for SITL

hendjoshsr71 opened the pull request:
This enables units tests to be run by CI when built under SITL. The reasoning for this came from PR #18365 & #18363 where I was trying to fix simulated drivers and spent far to long in the debugger when a unit test would have worked perfectly. 

It enables GTest to catch exit events such as an abort. 
Thus we can test that SITL creates Internal_Errors where required and prevent regressions as a bonus.

There are some details, such as the method to detect if the unit tests or running inside the `Panic()`, that I'm not sure will work across all platforms?

Log from the "test unit tests ...." tab in this PR showing the `Panic: AP_Internal_Error` being printed and caught by GTest for evaluation. 
![image](https://user-images.githubusercontent.com/69225461/130198335-5069c3cf-3443-4b4e-9e9b-604423f3c9d1.png)

hendjoshsr71 commented on libraries/AP_Common/tests/test_expandingstring_failure.cpp:
```diff
@@ -73,12 +73,28 @@ TEST(ExpandingString, Tests)
     EXPECT_EQ(0u, test_string->get_length());
-    EXPECT_FALSE(test_string->has_failed_allocation());
+
+// Odd change in behavior Why?? FALSE -> TRUE
+#if CONFIG_HAL_BOARD == HAL_BOARD_LINUX
+    // EXPECT_FALSE(test_string->has_failed_allocation());
+    EXPECT_TRUE(test_string->has_failed_allocation());
+#elif CONFIG_HAL_BOARD == HAL_BOARD_SITL
+    EXPECT_TRUE(test_string->has_failed_allocation());
+#endif
```
@khancyr I couldn't quite figure out the purpose and operation of these tests to figure out why this change of behavior was necessary to pass the tests?
 Same for below

hendjoshsr71 replied:
Fixed. The issue was due to the redefinition of realloc that was first being used by setenv() and getting count off by 1.

khancyr commented on tests/AP_gtest.h:
```diff
@@ -14,9 +15,17 @@ ::std::ostream& operator<<(::std::ostream& os, const class_name_& param) \
 #define AP_GTEST_MAIN() \
 int main(int argc, char *argv[]) \
 { \
+    setenv("SITL_PANIC_UNIT_TEST_EXIT", "1", 0); \
```
we should probably set those with WAF in the coverage setting

hendjoshsr71 replied:
@khancyr 
Would you point me in the right direction with an example? I've tried playing with the waf script to do this but I can't get it to work using cfg.setenv() 

I want to be sure that the tests can be run locally as well as on the server.

hendjoshsr71 replied:
Switched to using WEAK symbols instead

hendjoshsr71 commented on libraries/AP_HAL_SITL/system.cpp:
```diff
@@ -39,10 +39,21 @@ void panic(const char *errormsg, ...)
+    if (getenv("SITL_PANIC_EXIT") && !getenv("SITL_PANIC_UNIT_TEST_EXIT")) {
         // this is used on the autotest server to prevent us waiting
         // 10 hours for a timeout
-        exit(1);
```
Ekk this got deleted by accident. Will fix.

hendjoshsr71 commented:
@peterbarker 

I switched this to use `WEAK` symbols now and removed the function you weren't happy about. 
I do think like Buzz said on the call it may make sense to just build and do unit tests against SITL alone instead of Linux?

khancyr reviewed (approved):
LGTM
Reduce a bit the coverage stats but it will allow to cover the missing functions, so that is great!

I think that having the weak symbol on panic for SITL and Linux is fine

hendjoshsr71 commented:
> Reduce a bit the coverage stats but it will allow to cover the missing functions, so that is great!

@khancyr Could you explain to me how it reduces the coverage stats? I'm learning about coverage in general  so I'm interested to figure this out.

khancyr commented:
As we are now using SITL on coverage, we are passing over lines that were restricted to SITL only before. Those are mostly some panic and debug lines. But we don't necessary check them as there weren't there on Linus build. Therefore we test more stuff but check less outputs so our coverage statistic is dropping.

You can see it easily on the coverage report

hendjoshsr71 commented:
Ahh makes sense. So you are telling me I need to add some more tests to get your coverage stats back up ;)

I ran the coverage report for the only test that is SITL only SIM_MS5611.  They were quite useful to find that one of my tests in that wasn't testing a few lines they were supposed to be.

peterbarker reviewed (approved):
This looks fine to me.

My understanding is this would've helped with some testing that was being done as part of some sim refactoring.

@khancyr what do you think?

khancyr commented:
I have tested and it is working. So that is fine with me.
